src/transformations.rs

```diff
@@ -368,14 +380,63 @@ impl Closurizable for RichTerm {
     /// Generate a fresh variable, bind it to the corresponding closure `(t,with_env)` in `env`,
     /// and return this variable as a fresh term.
     fn closurize(self, env: &mut Environment, with_env: Environment) -> RichTerm {
+        // If the term is already a variable, we don't have to create a useless intermediate
+        // closure. We just transfer the original thunk to the new environment. This is not only an
+        // optimization: this is relied upon by recursive record merging when computing the
+        // fixpoint.
+        //
+        // More specifically, the evaluation of a recursive record patches the environment of each
+        // field with the thunks recursively referring to the other fields of the record. `eval`
+        // assumes that a recursive record field is either a constant or a generated variable whose
+        // thunks *immediately* contain the original unevaluated expression (both properties are
+        // true after the share normal form transformation and maintained when reverting thunks
+        // before merging recursive records).
+        //
+        // To maintain this invariant, `closurize` must NOT introduce an indirection through a
+        // variable, such as transforming:
+        //
+        // ```
+        // {foo = %1, bar = 1} in env %1 <- 1 + bar
+        // ```
+        //
+        // to:
+        //
+        // ```
+        // {foo = %2, bar = 1} in env %2 <- (%1 in env %1 <- 1 + bar)
+        // ```
+        //
+        // In this case, the evaluation of the recursive records will patch the outer environment
+        // instead of the inner one, giving:
+        //
+        // ```
+        // {foo = %2, bar = 1} in env %2 <- (%1 in env %1 <- 1 + bar), bar <- 1
+        // ```
+        //
+        // Then, evaluating `foo` would unduly raise an unbound identifier error.
+        //
+        //
+        // We currently only do this optimization for generated variables (introduced by the share
+        // normal form). We could do it for non-generated identifiers as well, but be we would be
+        // renaming user-supplied variables by gibberish generated names. It may hamper error
+        // reporting, so for the time being, we restrict ourselves to generated identifiers.
```

Comment:
And it will never break the recursive record merging to not handle user-defined variables this way?

Reply:
TL;DR Theoretically it should be fine for closurize to handle user-defined variables either way, but you made me realize the share normal form may be breaking the invariant in some cases, which might break things if we were to optimize user-defined variables as well at some point.

Very good question. After thinking a bit more about it, I think it is actually working whether we do the optimization for user-supplied variable or not, but somehow by chance, while it should work by design.

The invariant merging is relying on is that every record field is a generated variable that itself contains the original expression (excepted for constants, that we don't really care about). Because this is a generated variable, the fact that we do or don't do the optimization for non-generated variable shouldn't matter.

But, you just made me realize this invariant may not exactly be always true. Because the share normal form transformation also tries not to introduce useless indirection, it doesn't do anything for fields that are already containing just a variable, such as `{foo = bar, bar = 1}`. There, the share normal form transformation would leave the record as it is. This violates the invariant, and the following example should intuitively not work as intended:

```
({ foo | default = 1, bar = foo} & {}).foo
```

However, it does work. I suspect it is precisely *because we do add an indirection* in `closurize` for non-generated variables currently, causing `closurize` to somehow make up for the behavior of the share normal form transformation. But by accident, and if we were to extend of the optimization to all variables, that might break the example above.

I'll try out a few things to validate those hypotheses, but if it's the case, I imagine the simplest is to remove this optimization from the share normal form so that it does enforce the sought invariant at the cost of adding a superfluous indirection for fields that contains a variable.

Reply:
False alert, share normal form was already doing the right thing, actually, always introducing an indirection. I'll add a few corresponding tests, just in case.

Reply:
So, maybe you need to adjust your comment a little before merging.